Sources/PackageCollections/PackageIndex.swift

```diff
@@ -0,0 +1,186 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2022 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import Dispatch
+import Foundation
+import PackageModel
+import TSCBasic
+
+struct PackageIndex: PackageIndexProtocol {
+    private let configurationStorage: PackageIndexConfigurationStorage
+    private let httpClient: HTTPClient
+    private let callbackQueue: DispatchQueue
+    private let observabilityScope: ObservabilityScope
+    
+    // Feature flag: env var SWIFTPM_ENABLE_PACKAGE_INDEX
+    let enabled: Bool
+
+    // TODO: cache metadata results
+
+    init(
+        fileSystem: FileSystem,
+        customHTTPClient: HTTPClient? = nil,
+        callbackQueue: DispatchQueue,
+        observabilityScope: ObservabilityScope
+    ) {
+        self.configurationStorage = PackageIndexConfigurationStorage(fileSystem: fileSystem)
+        self.httpClient = customHTTPClient ?? HTTPClient()
+        self.callbackQueue = callbackQueue
+        self.observabilityScope = observabilityScope
+        self.enabled = ProcessInfo.processInfo.environment["SWIFTPM_ENABLE_PACKAGE_INDEX"] == "1"
+    }
+
+    func set(
+        url: Foundation.URL,
+        callback: @escaping (Result<Void, Error>) -> Void
+    ) {
+        let callback = self.makeAsync(callback)
+        
+        guard self.enabled else {
+            return callback(.failure(PackageIndexError.featureDisabled))
+        }
+        
+        self.configurationStorage.update(index: .init(url: url)) { result in
+            callback(result.tryMap { () })
+        }
+    }
+
+    func unset(callback: @escaping (Result<Void, Error>) -> Void) {
+        let callback = self.makeAsync(callback)
+        
+        guard self.enabled else {
+            return callback(.failure(PackageIndexError.featureDisabled))
+        }
+        
+        self.configurationStorage.update(index: .init(url: nil)) { result in
+            callback(result.tryMap { () })
+        }
+    }
```

Comment:
not sure if we need the set/unset to be part of the API for PackageIndex. in other places in SwiftPM similar business logic modules get a readonly copy of the configuration in the initializer so they can read the configuration. the CLI / Xcode deal with set/unset using the "storage" type data structures. in this case this means the CLI deals with PackageIndexConfigurationStorage, and PackageIndex initializer would take PackageIndexConfiguration which is a readonly representation of the active configuration

Reply:
This means package index URL doesn't change until next time it gets initialized, correct?

Reply:
as an example you see the mirrors passed into workspace, and the mirrors storage which is used to set/unset that is only used by the CLI (and IDEs)

Reply:
> This means package index URL doesn't change until next time it gets initialized, correct?

yes, which is think is fine imo. the CLI will initialize this on every call anyways, and the Xcode integration layer could do the same if it needs to

Reply:
amended https://github.com/apple/swift-package-manager/pull/4010/commits/c1d364a022d867b5bfcd9f69006a1fc714001414